src/app/core/services/upload-attachments/upload-attachments.service.ts

```diff
@@ -0,0 +1,143 @@
+  providedIn: 'root'
+})
+export class UploadAttachmentsService {
+
+  constructor(
+    private httpClient: HttpClient,
+    private progressService: ProgressService,
+    private newAttachmentService: NewAttachmentService,
+    private apiAttachmentService: ApiAttachmentService,
+    private loggingService: LoggingService
+  ) {}
+
+  /**
+   * Upload attachments
+   *
+   * When we upload attachments, we get an attachment id back. The attachment ids are later used to "join" the
+   * uploaded attachments to the registration.
+   *
+   * After attachments are uploaded, we request an update of local attachment metadata so that we do not need to update
+   * the attachments again later if the post/put registration request fails.
+   *
+   * @returns Attachments with attachment ids from upload response
+   * @throws {UploadAttachmentError} If the attachment upload fails
+   */
+  async uploadAllAttachments(draft: RegistrationDraft): Promise<AttachmentUploadEditModel[]> {
+    const attachments = await firstValueFrom(this.newAttachmentService.getAttachments(draft.uuid));
+
+    // Some attachments may already be uploaded
+    const alreadyUploaded = attachments.filter(a => a.AttachmentUploadId != null);
+    const attachmentsToUpload = attachments.filter(a => a.AttachmentUploadId == null);
+
+    // Error handling
+    // wrap this.uploadAttachment in a function that saves exceptions so that we can handle those that fail later
+    const failedAttachmentIds = [];
+    const uploadAttachmentAndHandleErrors = async (attachment: AttachmentUploadEditModel) => {
+      try {
+        return await this.uploadAttachment(attachment, draft);
+      } catch (error) {
+        this.loggingService.error(error, DEBUG_TAG, 'Failed to upload attachment', attachment.AttachmentId);
+        failedAttachmentIds.push(attachment.AttachmentId);
+      }
+    };
+
+    // Upload all attachments concurrently
+    const uploadedAttachments = await Promise.all(attachmentsToUpload.map(uploadAttachmentAndHandleErrors));
+
+    if (failedAttachmentIds.length) {
+      throw new UploadAttachmentError(draft.uuid, failedAttachmentIds);
+    }
+
+    return [
+      ...alreadyUploaded,
+      ...uploadedAttachments
+    ];
+  }
+
+  private onHttpEvent(event: HttpEvent<any>, attachment: AttachmentUploadEditModel) {
+    this.loggingService.debug('Attachment upload http event', DEBUG_TAG, event);
+    if (event.type === HttpEventType.UploadProgress) {
+      this.progressService.setAttachmentProgress(attachment.id, event.total, event.loaded);
+    }
+  }
+
+  private onHttpResponseEvent(event: HttpResponse<string>) {
+    if (event instanceof HttpErrorResponse) {
+      // This is already an error and contains useful info, so we can just throw it
+      throw event;
+    }
+    if (!event.ok) {
+      throw new Error(`Could not upload attachment: Status: ${event.statusText}`);
+    }
+    return event.body;
+  }
+
+  private sendPostRequestWithImageBlob(blob: Blob) {
```

Comment:
Pirk: Skal det være en returtype her?

Reply:
Tenker det går greit på private metoder, særlig når man ikke får noe ytterligere hjelp fra typescript, det spesifiseres allerede at det er en `Observable<HttpEvent<string>>` i bunnen av funksjonen. Men jeg kan godt legge det til om du syns det er best å ha med begge steder.

Reply:
Åkei, ikke viktig for meg